src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -330,20 +319,34 @@ private void EventLoop()
                     // The native shim is responsible for ensuring this condition.
                     Debug.Assert(numEvents > 0, $"Unexpected numEvents: {numEvents}");
 
-                    for (int i = 0; i < numEvents; i++)
+                    // Look up contexts first to improve locality and reduce number of lock operations.
+                    lock (s_lock) // for _handleToContextMap
                     {
-                        IntPtr handle = _buffer[i].Data;
-                        if (handle == ShutdownHandle)
-                        {
-                            shutdown = true;
-                        }
-                        else
+                        for (int i = 0; i < numEvents; i++)
                         {
-                            SocketAsyncContext context = GetContextFromHandle(handle);
-                            if (context != null)
+                            IntPtr handle = _buffer[i].Data;
+                            if (handle == ShutdownHandle)
                             {
-                                context.HandleEvents(_buffer[i].Events);
+                                shutdown = true;
                             }
+                            else
+                            {
+                                Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
+                                SocketAsyncContext context;
+                                _handleToContextMap.TryGetValue(handle, out context);
+                                contexts[i] = context;
```

Comment:
`_handleToContextMap.TryGetValue(handle, out contexts[i]);`?

Reply:
And since shutdown should be relatively rare, maybe:
```C#
IntPtr handle = _buffer[i].Data;
Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
_handleToContextMap.TryGetValue(handle, out contexts[i]);
shutdown |= (handle == ShutdownHandle);
```
?  I don't know if that would have any benefit; since it should be rare, presumably branch prediction would do well here, anyway.

Reply:
I'm surprised that is possible. :astonished: Do we reference the item in the array? Or does the compiler transform this to something similar like what is written.

Reply:
> Do we reference the item in the array? 

Yes... it passes the array slot by ref, see the ldelema here:
https://sharplab.io/#v2:EYLgHgbALANALiAlgGwD4AEAMACdBGAbgFgAoLXPAOgCUBXAOzkQFsBTSgYQHtmAHFVgCcAykIBuiAMasAzsRKl0AZlwAmbB2wBvUtj24V6KNgCyACkSMA2gF1sAQ0GCY2S3FcBKbbv2+AQo5mgqwAZg5OVog2HvK+AL4+eonYViascAAWXAAmAJJ8yGZpmTn5vMgA8rxMXPQylAByXLn0yJaWAOYeNsnKuMYBgkGhrozYMshccB7JOiTxpHFAA=
and then the lea here:
https://sharplab.io/#v2:EYLgHgbALANALiAhgZwLYB8ACAGABJgRgG4BYAKB3wIDoAlAVwDs4BLVAU2oGEB7VABxYAbdgCcAymIBuLAMbtkpMuUwBmfACZcXXAG9yuQ/nWYouALIAKFswDaAXVyJRomLhtx3ASj0Gj/gCFnS1F2ADMnF1sWey8lfwBfP0Nk3FtzdjgACx4AEwBJASFLDOy8wv4hAHl+Vh5GZGoAOR58xiEbGwBzL3tUtXwzINEQ8PdmXGQhHjgvVP0yRPIEoA===